pkg/metrics/server.go

```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
 		Handler: router,
 	}
 
-	go func() {
-		if err := srv.ListenAndServeTLS(tlsCRT, tlsKey); err != nil {
-			klog.Errorf("error starting metrics server: %v", err)
-		}
-	}()
-	<-stopCh
-	if err := srv.Shutdown(context.Background()); err != nil && err != http.ErrServerClosed {
-		klog.Errorf("error stopping metrics listener: %v", err)
+	return srv
+}
+
+func startServer(srv *http.Server, stopCh <-chan struct{}) {
+	klog.Infof("starting metrics server")
+	if err := srv.ListenAndServeTLS(tlsCert, tlsKey); err != nil && err != http.ErrServerClosed {
+		klog.Errorf("error from metrics server: %v", err)
 	}
 }
 
-// PodLabelsUsed indicates whether the deprecated Pod label matching functionality
-// is turned on.
-func PodLabelsUsed(enable bool) {
-	if enable {
-		podLabelsUsed.Set(1)
+func stopServer(srv *http.Server) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := srv.Shutdown(ctx); err != nil {
+		klog.Errorf("error or timeout stopping metrics listener: %v", err)
+	}
+	klog.V(2).Infof("stopping metrics server without error")
+}
+
+// RunServer starts the server, and watches the tlsCert and tlsKey for certificate changes.
+func RunServer(port int, stopCh <-chan struct{}) {
+	srv := buildServer(port)
+	if srv == nil {
 		return
 	}
-	podLabelsUsed.Set(0)
+
+	go startServer(srv, stopCh)
+
+	// Set up and start the file watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Errorf("failed to create watcher: %v", err)
+	}
+	defer watcher.Close()
+	if err := watcher.Add(certsDir); err != nil {
+		klog.Errorf("failed to add %v to watcher, cert/key rotation will be disabled: %v", certsDir, err)
+	}
+
+	origCertChecksum := checksumFile(tlsCert)
+	origKeyChecksum := checksumFile(tlsKey)
+
+	for {
+		select {
+		case <-stopCh:
+			stopServer(srv)
+		case event, ok := <-watcher.Events:
+			if !ok {
```

Comment:
I wonder if we even need this check.  I have been happily runing with the detection of channel closing in `tuned` for years.  Also, my advice of add "cert/key rotation will be disabled" is probably not a good one looking more closely.  My appologies.  This might be just a temporary blip and the rotation code might be completely unaffected.  `watcher.Add()` are fatal though in my view and not checking for `nil` will bring down the whole operator so we need to be extra careful there.

Reply:
I removed this check. If somehow `event` were set to nil here, I believe the worst case is that we would check if the files really changed.